### ISSUE TITLE
Update bank statement validation

### ISSUE THREAD

**Issue by cesarcosta99** [woocommerce/woocommerce-gateway-stripe#2189]
## Description

Let's make sure we apply the same validation rules [determined by Stripe](https://stripe.com/docs/statement-descriptors#requirements) to both full and short bank statement fields.

While working on this, let's also consider validating against the list of blocked strings (see pc4etw-bL-p2 and the referenced issue below).

The validation error for that setting:
>[…] should be between 5 and %2$u characters long, contain at least single Latin character and **does** not contain special characters […]

could be adjusted to:
>[...] should be between 5 and %2$u characters long, contain at least a single Latin character, and not contain any of the following special characters: […]

## Acceptance criteria

- Validation rules are applied in both full and shortened bank statement fields.
- Validation rules match Stripe requirements.
- The error message has been adjusted.
- The blocked strings validation is implemented or there's a reason why we should not implement it.

## Next steps

**Clean statement descriptor fields before saving them**

We should consider to make use of the [clean_statement_descriptor](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-helper.php#L482) helper before saving the statement descriptor fields in the settings. That would increase confidence that the data would not make the intents fail and we would not need to clean every time we send them.

The current method may require refactoring considering the new validation requirements and it's quite old. If you decide not to tackle this while working on this issue, please create a follow up issue.

**Comment by cesarcosta99** [comment]
Blocked due to p1638293368433600/1637246174.324800-slack-C01BZUL57SQ.

**Comment by ricardo** [comment]
> An extra mile for this task would be to test and confirm the value we set is sent over to the payment intent.

When UPE is enabled, the value is not passed to the payment intent. This is a bug, perhaps unrelated to this issue, but seems easy to fix.

> Make sure we apply the same validation rules determined by Stripe to the full bank statement field

Unfortunately, it's not possible to validate the common terms requirement (Issue https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1627): "Contains more than a single common term or common website URL...".

More about this here: p1638293368433600/1637246174.324800-slack-C01BZUL57SQ

To sum up the next steps on this, I guess we have two options: (a) We can either use a custom statement descriptor without the common terms validation, or (b) remove the custom statement descriptor and use the one set from the Stripe dashboard.

In the (b) scenario, we would be removing from the current behavior and not allowing multiple websites use the same Stripe account with different statement descriptors. Detailed discussion is in the thread above.

**Comment by cesarcosta99** [comment]
As a result of the p1638293368433600/1637246174.324800-slack-C01BZUL57SQ discussion, we're going forward with option (a) Ricardo described. We need to make sure the validation rules are being applied to the statement descriptor fields, with exception to the common terms; and the error message is accurate.

>An extra mile for this task would be to test and confirm the value we set is sent over to the payment intent.

>If left blank, the transaction will use what's set on the Stripe account. An idea would be to get that value, since we request the Stripe account details, and set it as a placeholder for the field.

Not necessary to tackle both things above while working on this, I'll be addressing that on #2205.

---

Edit: @ricardo, perhaps we can now close #1627 since it seems to be impossible to address that on our end.

**Comment by ricardo** [comment]
@cesarcosta99, I closed https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1627.

**Comment by cesarcosta99** [comment]
It seems we can still consider to include the wording validation (#1627) in the scope of this issue. See https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1627#issuecomment-1012595461

**Comment by cesarcosta99** [comment]
Updated the description to reflect the removal of some requirements and include new details.

**Comment by a-danae** [comment]
👋 Trying to understand the desired behavior for descriptor requirements.

> Contains only Latin characters

Do we want to allow accentuated Latin characters? For example, `àaáâæäãå`.
When passing this to the descriptor, Stripe transforms it to `aaaaaaa`. We could allow them and let Stripe change them, or we could throw an error asking the merchant to update the descriptor.

> The blocked strings validation is implemented or there's a reason why we should not implement it.

Is this list still valid? I tried using some of the words mentioned there as the statement descriptor and they worked just fine. It was sent with the payment intent, the payment was processed just fine, and the descriptor was set in the Payment details under the Stripe dashboard.

**Comment by a-danae** [comment]
A note on this

> That would increase confidence that the data would not make the intents fail and we would not need to clean every time we send them

Looks like Stripe cleans the descriptor on its end so the payment intents don't fail. I bypassed the cleaning under `clean_statement_descriptor` and the validation under `validate_account_statement_descriptor_field` and the payment intents were processed just fine.

I set the statement descriptor to `jpmorgan?+_*-aה&جَدِ シ`, containing a word from the blocklist, allowed and disallowed special chars, and non-Latin characters. I then made a successful purchase with a 3DS card. In the Stripe dashboard, the descriptor value was set to `JPMORGAN?+_*-A&AMP;` for this payment.

So maybe this issue would be more to improve the UX (so the merchants have a consistent descriptor between the extension setting and the one used in the final bank statement) and not so much one that would affect the payment flow.

**Comment by a-danae** [comment]
- Opened a PR for one of the Stripe's requirements - https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2682

And created two issues for the other behaviors mentioned in this one. Having them separately will make them easier to test and review:
- https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2684 
- https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2685 (needs feedback)

**Comment by cesarcosta99** [comment]
@a-danae it's been a long time I touched this so I took some time to catch-up.

>When passing this to the descriptor, Stripe transforms it to aaaaaaa. We could allow them and let Stripe change them, or we could throw an error asking the merchant to update the descriptor.

I took a quick glance on your PR and, IMO, you went down with the best approach, which is throwing an error asking the merchant to change it 👍 

>Is this list still valid? I tried using some of the words mentioned there as the statement descriptor and they worked just fine.
>Looks like Stripe cleans the descriptor on its end so the payment intents don't fail. I bypassed the cleaning under clean_statement_descriptor and the validation under validate_account_statement_descriptor_field and the payment intents were processed just fine.

Thank you for getting this moving forward by checking it in the post. I might be wrong, but I think the blocked words might have worked out in a test scenario but the chances of a PI fail should be higher in a real case scenario. In the end, that's up to Stripe, but we can try to keep up with them to be consistent and tentatively avoid this.

Also, good job on breaking this in a couple of issues! Just one thing I'd like to emphasize is that in #2684 we'll likely be able to simplify the logic around this in the payment flow by removing this validation, once we'll validate it before saving it.

**Comment by a-danae** [comment]
Thanks a bunch for checking out the issue and replying back, @cesarcosta99!

> I might be wrong, but I think the blocked words might have worked out in a test scenario but the chances of a PI fail should be higher in a real case scenario. In the end, that's up to Stripe, but we can try to keep up with them to be consistent and tentatively avoid this.

Cool, handling it on our end sounds good.

> Just one thing I'd like to emphasize is that in https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2684 we'll likely be able to simplify the logic around this in the payment flow by removing this validation, once we'll validate it before saving it.

Checking if I got this. At the moment, we're calling [clean_statement_descriptor](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-helper.php#L576) every time we send a request to Stripe containing the bank descriptor. 
So, once we validate and clean the bank descriptor before saving it, we could remove the calls to clean_statment_descriptor we're currently making for the Stripe requests, correct?

**Comment by cesarcosta99** [comment]
>So, once we validate and clean the bank descriptor before saving it, we could remove the calls to clean_statment_descriptor we're currently making for the Stripe requests, correct?

That's right 💯
